modules/tinymce/src/plugins/advlist/main/ts/ui/Buttons.ts

```diff
@@ -26,13 +25,15 @@ const normalizeStyleValue = (styleValue: string | undefined): string =>
   Type.isNullable(styleValue) || styleValue === 'default' ? '' : styleValue;
 
 const makeSetupHandler = (editor: Editor, nodeName: ListType) => (api: Toolbar.ToolbarSplitButtonInstanceApi | Toolbar.ToolbarToggleButtonInstanceApi) => {
-  const nodeChangeHandler = (e: EditorEvent<NodeChangeEvent>) => {
-    api.setActive(ListUtils.inList(editor, e.parents, nodeName));
-    api.setEnabled(!ListUtils.isWithinNonEditableList(editor, e.element));
+  const updateButtonState = (element: Element, parents: Node[]) => {
+    api.setActive(ListUtils.inList(editor, parents, nodeName));
+    api.setEnabled(!ListUtils.isWithinNonEditableList(editor, element));
   };
-  editor.on('NodeChange', nodeChangeHandler);
+  const nodeChangeHandler = (e: NodeChangeEvent) => updateButtonState(e.element, e.parents);
+  const element = editor.selection.getStart(true);
+  updateButtonState(element, editor.dom.getParents(element));
```

Comment:
Why do we need to call the `updateButtonState` function explicitly here? It will be called one more time right after this call from `ListUtils.setNodeChangeHandler(editor, nodeChangeHandler)`. This call feels a bit redundant.

Reply:
Or the `ListUtils.setNodeChangeHandler`. Unless I'm missing something, the `updateButtonState` is unnecessary called twice.

Reply:
No, you're right. The new code calls it twice - using different inputs (`editor.selection.getStart(true)` and then `editor.selection.getNode()`)

I think inside `setNodeChangeHandler` is the correct place, but it should use `getStart` as spocke said in the other thread.

- [ ] only set the initial state once

Reply:
You don't need the `setNodeChangeHandler` since line 36 should just be. `return () => editor.off('NodeChange', nodeChangeHandler);` but you could pull this apart into that module if you like but still use the non fake event approach.

Reply:
What about something like this:
```javascript
const updateButtonState = (editor: Editor, parents: Node[]) => {
    const element = editor.selection.getStart(true);
    api.setActive(ListUtils.inList(editor, parents, nodeName));
    api.setEnabled(!ListUtils.isWithinNonEditableList(editor, element));
  };
  const nodeChangeHandler = (e: NodeChangeEvent) => updateButtonState(editor, e.parents);

  return ListUtils.setNodeChangeHandler(editor, nodeChangeHandler);
```
it does not use the element in the fake event and also does not call `updateButtonState`

Reply:
Yes the easy fix is to just get rid of the extra `updateButtonState(element, editor.dom.getParents(element));` and keep the separate function. As suggested.

Reply:
done, let me know if it is ok